src/cephadm/cephadmlib/runscripts.py

```diff
@@ -0,0 +1,235 @@
+import contextlib
+import json
+import pathlib
+import shlex
+
+from typing import Any, Dict, Union, List, IO, TextIO, Optional, cast
+
+from .container_engines import Podman
+from .container_types import CephContainer, InitContainer
+from .context import CephadmContext
+from .context_getters import fetch_meta
+from .daemon_identity import DaemonIdentity
+from .file_utils import write_new
+from .net_utils import EndPoint
+
+
+# Ideally, all ContainerCommands would be converted to init containers. Until
+# that is done one can wrap a CephContainer in a ContainerCommand object and
+# pass that as a pre- or post- command to run arbitrary container based
+# commands in the script.
+class ContainerCommand:
+    def __init__(
+        self,
+        container: CephContainer,
+        comment: str = '',
+        background: bool = False,
+    ):
+        self.container = container
+        self.comment = comment
+        self.background = background
+
+
+Command = Union[List[str], str, ContainerCommand]
+
+
+def write_service_scripts(
+    ctx: CephadmContext,
+    ident: DaemonIdentity,
+    *,
+    container: CephContainer,
+    init_containers: Optional[List[InitContainer]] = None,
+    endpoints: Optional[List[EndPoint]] = None,
+    pre_start_commands: Optional[List[Command]] = None,
+    post_stop_commands: Optional[List[Command]] = None,
+    timeout: Optional[int] = None,
+) -> None:
+    """Write the scripts that systemd services will call in order to
+    start/stop/etc components of a cephadm managed daemon. Also writes some
+    metadata about the service getting deployed.
+    """
+    data_dir = pathlib.Path(ident.data_dir(ctx.data_dir))
+    run_file_path = data_dir / 'unit.run'
+    meta_file_path = data_dir / 'unit.meta'
+    post_stop_file_path = data_dir / 'unit.poststop'
+    stop_file_path = data_dir / 'unit.stop'
+    image_file_path = data_dir / 'unit.image'
+    # use an ExitStack to make writing the files an all-or-nothing affair. If
+    # any file fails to write then the write_new'd file will not get renamed
+    # into place
+    with contextlib.ExitStack() as estack:
+        # write out the main file to run (start) a service
+        runf = estack.enter_context(write_new(run_file_path))
+        runf.write('set -e\n')
+        for command in pre_start_commands or []:
+            _write_command(ctx, runf, command)
+        init_containers = init_containers or []
+        if init_containers:
+            _write_init_container_cmds_clean(ctx, runf, init_containers[0])
```

Comment:
I'm a bit confused why we're writing a clean command for the first thing in the init_containers list. It looks like `_write_init_container_cmds` calls `_write_init_container_cmds_clean` for the given init container, and we're calling `_write_init_container_cmds` on every init container, including the first one, right after this. So what's the need for this extra call?

Reply:
This is also just a "move" of the code that implements init containers from before. But since I wrote it this year I don't have a problem with explaining or justifying it :-)

Reply:
Why do we only need the cleanup command for the first init container at the start though? If say, the second init container is the one that failed, will the cleanup command built from `init_containers[0]` clean that one up?

Reply:
Ah, good catch. _Originally_, because init containers always run sequentially the container name was going to be reused. So if service foo.bar had container name `ceph-UUID-foo-bar` all init containers would be named `ceph-UUID-foo-bar-init` and thus it was safe to always clean once on script start up.

So you've identified a real issue. It's not new to this PR because this part of the code is just getting moved. There are some other bugs I found with regards to init containers while working on smb. I think it would be easier for me to try and fix some of them together in PR #54722 - what do you think about that?

Reply:
I think I broke this when adding DaemonIdentity + DaemonSubIdentity FWIW.

Reply:
This has been updated as part of #54722